Pulling changes proposed in another fork by @SebaDro

brudo opened the pull request:
An upcoming project would have similar requirements for authenticated access to an MQTT hub. Not sure how quickly these changes will be merged to the upstream master, so I thought I would pull them from @SebaDro's fork independently to ensure efforts are not duplicated and to avoid conflicts.

Not sure if I will need the GeoEvent aware changes to the outbound transport to allow for template expansion against GeoEvent fields in the topic name, although I agree it could be useful. On initial review, I think there may be a potential reentrancy issue issue with the current implementation, which would manifest if different calls to `receive` are processed in parallel on different threads. (Perhaps declare the local `topic` variable outside of the `if` block, so that the call to `mqttClient.publish` can reference the local variable, and do not modify `this.topic`? That might resolve it.)

I am more interested in the inbound transport anyway, but will probably merge at least the authentication enhancement for both inbound and outbound directions.

SebaDro committed:
Change outbound transport to a GeoEvent aware outbound transport

Change outbound transport to a GeoEvent aware outbound transport to use
GeoEvent field values for the MQTT topic property.

SebaDro committed:
Connect to a MQTT broker with user credentials.

Add the capability for the input and output transport to connect with a
username and a password to a MQTT broker

brudo reviewed (commented):
On initial review, I think there may be a potential reentrancy issue issue with the current implementation, which would manifest if different calls to `receive` are processed in parallel on different threads. (Perhaps declare the local `topic` variable outside of the `if` block, so that the call to `mqttClient.publish` can reference the local variable, and do not modify `this.topic`? That might resolve it.)

brudo reviewed (commented):
Please revert contact name to the original dummy value yourname@yourcompany.com for merging purposes.

SebaDro commented:
You're right. There could be a problem when modifying `this.topic` inside `receive`. I will fix your remarks as soon as possible.

The idea behind the GeoEvent aware output was to notify a specific vehicle, if it has entered a GeoFence. E.g. you have several autonomous driving vehicles that publish there sensor data and positions. Now you want to determine, if some of these vehicle entered a GeoFence to notify them. 
Without a GeoEvent aware output you only have a general topic, for which all vehicles have to subscribe. So each vehicle has to look into the MQTT message to determine, if the warning was adressed to it. 
With the use of the GeoEvent aware output, you can append the ID of a vehicle to the topic, if it has entered a GeoFence. Now each vehicle can subscribe for a topic that has it's ID as suffix, so that a vehicle only gets the warning that was adressed to it.

SebaDro committed:
Fix reentrancy issue in receive method

Fix reentrancy issue which might manifest when calling receive in
different threads and revert the contact name to dummy address

brudo reviewed (commented):
I meant to post this on my own pull request but inadvertently posted it in the pull request with Esri. Anyway I guess it applies to both. Copying here for continuity...

That should fix it. A few small refinements to suggest:

IMO it would be good to still have a member variable `this.topic`, initialized / validated / defaulted up-front as before. Then in `receive` it would collapse to `String topic = this.topic;` with no call to `getProperty` needed there.

Second, in the `if` block on the following line, this suggestion could be considered premature optimization but also more closely matches the original behavior in the case where topic name is static: perhaps skip the call to `formatString` if no field names are referenced, e.g. by adding `&& topic.contains("$")`.

What do you think?

brudo commented:
Thanks for your quick turn-around on my initial feedback @SebaDro! If you implement the above additional refinements, or make similar changes to minimize the differences in the code path when this added functionality is not being used, I think I will be ready to merge all of this to my fork.

After that I'm planning to add SSL support as well, and will be happy to share that back.

The scenario you described for sending notifications to the originating device based on processing is a great use case for the new functionality you implemented. With a very large fleet you might want to consider some combination of server-side grouping and device-side filtering so that the number of topics does not become excessive. Anyway it's great to be able to use the same connection that is already open (on the device side) so that a second, potentially heavier protocol is not needed for notifications.

The same capability could also be used for remote control applications, not to drive the vehicles of course, but e.g. to activate sensors or other equipment on the ground based on vehicles of a specific type approaching, since MQTT is such a lightweight protocol and so prevalent in the Internet of Things.

SebaDro commented:
I think your code refinement suggestions are appropriate. Unnecessary calls to `formatString` will be avoided and there will be only small differences to the original code. Checking if field names are referenced with `topic.contains("$")` seems also appropriate to me, since "$" will be treated specially in MQTT topics anyway. So I will adjust the code to your remarks.

SSL support would be very useful, especially in combination with the MQTT authentification capability. So I'm looking forward to use it when you have implemented that feature.

SebaDro committed:
Small code refinements

Small code refinements to avoid unnecessary method calls and to minimize
the difference in the code path
